core/app/models/concerns/spree/user_address_book.rb

```diff
@@ -18,7 +18,15 @@ def mark_default(user_address)
             end
             user_address.persisted? ? user_address.update!(default: true, archived: false) : user_address.default = true
           end
-          reset
+        end
+
+        def merge_one(user_address)
+          if(user_address.id)
+            # this happens if someone adds back an old address they had previously soft deleted
+            new_assoc = self.to_a.reject {|ua| ua.id == user_address.id} + [user_address]
```

Comment:
Yes. In #prepare_user_address we seem to have a different instance than the one in this association. So when we update the archived flag it wouldn't match the one inside here.